File: flutter_app/lib/models/player_pitcher.dart
class PlayerPitcher {
  final String name;
  final String team;
  final int games;
  final int wins;
  final int losses;
  final int saves;
  final int holds;
  final double ip;      // 이닝
  final int hits;       // 피안타
  final int hr;         // 피홈런
  final int bb;         // 볼넷
  final int hbp;        // 사구
  final int so;         // 삼진
  final int runs;       // 실점
  final int er;         // 자책점
  final double era;     // 평균자책점
  final double whip;    // WHIP
  final int season;

  const PlayerPitcher({
    required this.name,
    required this.team,
    required this.games,
    required this.wins,
    required this.losses,
    required this.saves,
    required this.holds,
    required this.ip,
    required this.hits,
    required this.hr,
    required this.bb,
    required this.hbp,
    required this.so,
    required this.runs,
    required this.er,
    required this.era,
    required this.whip,
    required this.season,
  });

  // 고급 지표
  double get k9 => ip > 0 ? 9 * so / ip : 0;
  double get bb9 => ip > 0 ? 9 * bb / ip : 0;
  double get hr9 => ip > 0 ? 9 * hr / ip : 0;
  double get kbb => bb > 0 ? so / bb : 0;

  // FIP 상수는 시즌별로 다르나, KBO 평균값 4.0 사용
  double get fip {
    if (ip <= 0) return 0;
    const fipConst = 3.20;
    return (13 * hr + 3 * (bb + hbp) - 2 * so) / ip + fipConst;
  }

  factory PlayerPitcher.fromMap(Map<String, dynamic> map) {
    return PlayerPitcher(
      name: map['name'] ?? '',
      team: map['team'] ?? '',
      games: map['games'] ?? 0,
      wins: map['wins'] ?? 0,
      losses: map['losses'] ?? 0,
      saves: map['saves'] ?? 0,
      holds: map['holds'] ?? 0,
      ip: (map['ip'] ?? 0).toDouble(),
      hits: map['hits'] ?? 0,
      hr: map['hr'] ?? 0,
      bb: map['bb'] ?? 0,
      hbp: map['hbp'] ?? 0,
      so: map['so'] ?? 0,
      runs: map['runs'] ?? 0,
      er: map['er'] ?? 0,
      era: (map['era'] ?? 0).toDouble(),
      whip: (map['whip'] ?? 0).toDouble(),
      season: map['season'] ?? 2024,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'team': team,
      'games': games,
      'wins': wins,
      'losses': losses,
      'saves': saves,
      'holds': holds,
      'ip': ip,
      'hits': hits,
      'hr': hr,
      'bb': bb,
      'hbp': hbp,
      'so': so,
      'runs': runs,
      'er': er,
      'era': era,
      'whip': whip,
      'season': season,
    };
  }
}

File: crawler/scrapers/player_pitcher.py
"""
선수 투수 기본 기록 크롤러
https://www.koreabaseball.com/Record/Player/PitcherBasic/BasicOld.aspx

실제 컬럼: 순위, 선수명, 팀명, ERA, G, CG, SHO, W, L, SV, HLD, WPCT, TBF, IP, H, HR, BB, HBP, SO, R, ER
WHIP / K/9 / FIP 는 직접 계산
IP 형식: "180 2/3" → 분수 변환 필요
"""

from .base_scraper import BaseScraper

URL = 'https://www.koreabaseball.com/Record/Player/PitcherBasic/BasicOld.aspx'

_IP_FRACTIONS = {'1/3': 0.333, '2/3': 0.667}


def _parse_ip(val: str) -> float:
    """이닝 파싱: "180 2/3" → 180.667"""
    val = str(val).strip()
    if ' ' in val:
        parts = val.split(' ', 1)
        try:
            full = int(parts[0])
            frac = _IP_FRACTIONS.get(parts[1].strip(), 0.0)
            return round(full + frac, 3)
        except (ValueError, IndexError):
            return 0.0
    try:
        return float(val)
    except ValueError:
        return 0.0


class PlayerPitcherScraper(BaseScraper):

    URL = URL

    def scrape(self, season: int) -> list[dict]:
        print(f'[투수] {season}시즌 크롤링 시작...')
        raw_rows = self._crawl_all_pages(season)
        result = [self._parse_row(row, season) for row in raw_rows if row.get('선수명')]
        print(f'[투수] 총 {len(result)}명 수집 완료')
        return result

    def _parse_row(self, row: dict, season: int) -> dict:
        ip  = _parse_ip(row.get('IP', '0'))
        h   = self._safe_int(row.get('H', '0'))
        hr  = self._safe_int(row.get('HR', '0'))
        bb  = self._safe_int(row.get('BB', '0'))
        hbp = self._safe_int(row.get('HBP', '0'))
        so  = self._safe_int(row.get('SO', '0'))
        er  = self._safe_int(row.get('ER', '0'))

        # WHIP 직접 계산 (페이지에 없음)
        whip = round((bb + h) / ip, 2) if ip > 0 else 0.0

        return {
            'name':   row.get('선수명', ''),
            'team':   row.get('팀명', ''),
            'games':  self._safe_int(row.get('G', '0')),
            'wins':   self._safe_int(row.get('W', '0')),
            'losses': self._safe_int(row.get('L', '0')),
            'saves':  self._safe_int(row.get('SV', '0')),
            'holds':  self._safe_int(row.get('HLD', '0')),
            'cg':     self._safe_int(row.get('CG', '0')),
            'sho':    self._safe_int(row.get('SHO', '0')),
            'tbf':    self._safe_int(row.get('TBF', '0')),
            'ip':     ip,
            'hits':   h,
            'hr':     hr,
            'bb':     bb,
            'hbp':    hbp,
            'so':     so,
            'runs':   self._safe_int(row.get('R', '0')),
            'er':     er,
            'era':    self._safe_float(row.get('ERA', '0')),
            'whip':   whip,
            'season': season,
        }
